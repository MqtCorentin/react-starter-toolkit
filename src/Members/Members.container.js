import React from 'react';
import PropTypes from 'prop-types';
import { LoaderModes } from '@axa-fr/react-toolkit-all';
import withEnhancedCustomFetch from 'shared/hoc/withCustomFetch';
import setLoaderMode from 'shared/helpers/setLoaderMode';
import { useNotifications } from 'shared/components/Notifications';
import Members from './Members';
import { useMembers } from './Members.hook';

export const MembersContext = React.createContext({ onchangeOrder: () => {}, sorting: {}, addNotification: () => {}, stateNotifications: [] });
const { Provider: MembersProvider } = MembersContext;

export const MembersEnhanced = ({ useMembersFn, fetch, ...rest }) => {
  const { anomaly, isLoading, members, onChangeOrder, sorting } = useMembersFn({
    fetchCustom: fetch,
  });

  const { addNotification, onDeleteNotification, stateNotifications } = useNotifications();

  return (
    <MembersProvider value={{ onChangeOrder, sorting, addNotification, stateNotifications }}>
      <Members
        {...rest}
        members={members}
        loaderMode={setLoaderMode({ isLoading, LoaderModes })}
        anomaly={anomaly}
        deleteNotification={onDeleteNotification}
        notifications={stateNotifications}
      />
    </MembersProvider>
  );
};

MembersEnhanced.propTypes = {
  useMembersFn: PropTypes.func,
};

MembersEnhanced.defaultProps = {
  useMembersFn: useMembers,
};

export default withEnhancedCustomFetch(MembersEnhanced);
