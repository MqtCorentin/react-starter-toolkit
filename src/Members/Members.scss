@import '@axa-fr/react-toolkit-core/dist/assets/scss/core.scss';
@import '../shared/scss/mixins.scss';

.af {
  &-btn {
    &__img {
      width: 19px;
      fill: $white;
      @include centerer;
    }
    &--circle-small {
      &:hover {
        .af {
          &-btn {
            &__img {
              fill: $color-axa;
            }
          }
        }
      }
    }
    &--circle-disabled {
      background-color: $color-gray-2;
      border-color: $color-gray-1;
      cursor: not-allowed;
      color: $color-gray-1;
      &:hover {
        background-color: $color-gray-2;
        border-color: $color-gray-1;
        color: $color-gray-1;
      }
    }
  }

  &-actions {
    display: flex;
    justify-content: space-evenly;
  }
}
