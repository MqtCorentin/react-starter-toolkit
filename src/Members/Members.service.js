export default ({ fetchCustom, signal, field = 'lastname', order = -1, max = 50, skip = 0 }) =>
  fetchCustom(`members?max=${max}&sort=${field}&dir=${order}&skip=${skip}`, { signal });
