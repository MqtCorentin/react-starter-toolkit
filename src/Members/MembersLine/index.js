export { default } from './MembersLine.container';
