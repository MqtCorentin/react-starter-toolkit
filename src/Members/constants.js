export const TITLE_BAR = 'Sociétaires';
export const TITLE = 'Liste des sociétaires';

export const FETCH_MEMBERS_INIT = 'FETCH_MEMBERS_INIT';
export const FETCH_MEMBERS_SUCCESS = 'FETCH_MEMBERS_SUCCESS';
export const FETCH_MEMBERS_FAILURE = 'FETCH_MEMBERS_FAILURE';
export const FETCH_MEMBERS_ORDER = 'FETCH_MEMBERS_ORDER';

export const TABLE_HEADERS_MEMBERS = [
  { label: 'Prénom', field: 'firstname' },
  { label: 'Nom', field: 'lastname' },
  { label: 'Date de naissance', field: 'birthdate' },
  { label: 'Sexe', field: 'sexe' },
];
