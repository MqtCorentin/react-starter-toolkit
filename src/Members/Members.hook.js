import { useEffect, useReducer, useCallback } from 'react';
import moment from 'moment';
import { get } from 'lodash';
import { NONE } from 'shared/constants';
import fetchData from 'shared/helpers/fetchData';
import findMembers from './Members.service';
import { FETCH_MEMBERS } from './constants';

const initState = {
  isLoading: false,
  members: [],
  anomaly: null,
  sorting: {
    field: '',
    order: NONE,
  },
  filters: {
    numberItems: 10,
    currentPage: 1,
    numberPages: 1,
  },
};

export const computeInfos = ({ members = [], getFn = get, momentFn = moment }) =>
  members.map(member => ({
    ...member,
    birthdate: momentFn(getFn(member, 'birthdate')).format('DD/MM/YYYY'),
  }));

export const setStateMembersLoading = ({ state }) => ({
  ...state,
  isLoading: true,
});

export const setStateMembersSuccess = ({ state, payload, computeInfosFn = computeInfos }) => ({
  ...state,
  isLoading: false,
  members: computeInfosFn({
    members: payload.members.responseBody,
  }),
});

export const setStateMembersFailure = ({ state, payload }) => ({
  ...state,
  isLoading: false,
  anomaly: payload.members,
});

export const setStateMembersOrder = ({ state, payload }) => ({
  ...state,
  sorting: {
    ...state.sorting,
    ...payload,
  },
  isLoading: true,
});

export const dataFetchReducer = (
  state,
  {
    type,
    payload,
    setStateMembersLoadingFn = setStateMembersLoading,
    setStateMembersSuccessFn = setStateMembersSuccess,
    setStateMembersFailureFn = setStateMembersFailure,
    setStateMembersOrderFn = setStateMembersOrder,
  },
) => {
  switch (type) {
    case FETCH_MEMBERS.INIT:
      return setStateMembersLoadingFn({ state });
    case FETCH_MEMBERS.SUCCESS:
      return setStateMembersSuccessFn({ state, payload });
    case FETCH_MEMBERS.ORDER:
      return setStateMembersOrderFn({ state, payload });
    case FETCH_MEMBERS.FAILURE:
      return setStateMembersFailureFn({ state, payload });
    default:
      return state;
  }
};

export const setMembersInit = dispatch => () => dispatch({ type: FETCH_MEMBERS.INIT });
export const setMembersError = dispatch => payload => dispatch({ type: FETCH_MEMBERS.FAILURE, payload });
export const setMembersSuccess = dispatch => payload => dispatch({ type: FETCH_MEMBERS.SUCCESS, payload });
export const setMembersOrder = dispatch => payload => dispatch({ type: FETCH_MEMBERS.ORDER, payload });

export const useMembers = ({
  fetchCustom,
  initStateCt = initState,
  fetchDataFn = fetchData,
  dataFetchReducerFn = dataFetchReducer,
  findMembersFn = findMembers,
  setMembersInitFn = setMembersInit,
  setMembersErrorFn = setMembersError,
  setMembersSuccessFn = setMembersSuccess,
  setMembersOrderFn = setMembersOrder,
}) => {
  const [stateMembers, dispatch] = useReducer(dataFetchReducerFn, initStateCt);
  const {
    sorting: { field, order },
  } = stateMembers;

  const onChangeOrder = useCallback(sorting => setMembersOrderFn(dispatch)(sorting), [setMembersOrderFn]);

  useEffect(() => {
    fetchDataFn({
      fetchCustom,
      setInit: setMembersInitFn(dispatch),
      setError: setMembersErrorFn(dispatch),
      setSuccess: setMembersSuccessFn(dispatch),
      fetchServices: {
        members: {
          service: findMembersFn,
          args: {
            field,
            order,
          },
        },
      },
    });
  }, [fetchCustom, fetchDataFn, field, findMembersFn, order, setMembersErrorFn, setMembersInitFn, setMembersSuccessFn]);

  return { ...stateMembers, onChangeOrder };
};
