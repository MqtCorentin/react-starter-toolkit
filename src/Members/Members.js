import React from 'react';
import Resilience from 'shared/components/Resilience/Resilience';
import { Paging } from '@axa-fr/react-toolkit-all';
import Loader from 'shared/components/Loader';
import Table from 'shared/components/Table';
import { TITLE_BAR, TITLE, TABLE_HEADERS_MEMBERS } from './constants';

const Members = ({ header, footer, title, menu, loaderMode, members, anomaly, onChangePaging, pagination, onChangeSorting, sorting }) => (
  <>
    {header()}
    {menu()}
    {title({ title: TITLE_BAR, backHome: true })}
    <div className="af-main container">
      <h1 className="af-title--content">{TITLE}</h1>
      <Loader mode={loaderMode}>
        <Resilience anomaly={anomaly}>
          <Table items={members} headers={TABLE_HEADERS_MEMBERS} onSort={onChangeSorting} sorting={sorting} />
          <Paging {...pagination} onChange={onChangePaging} id="paging" />
        </Resilience>
      </Loader>
    </div>
    {footer()}
  </>
);

export default Members;
