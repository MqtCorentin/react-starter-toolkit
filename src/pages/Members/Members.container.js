import React, { createContext } from 'react';
import { func } from 'prop-types';
import setLoaderMode from 'shared/helpers/setLoaderMode';
import Members from './Members';
import { useMembers } from './Members.hook';
import { SERVICE_NAME } from './constants';

export const MembersContext = createContext({ onChangeSorting: null, sorting: {} });
const { Provider: MembersProvider } = MembersContext;

const MembersEnhanced = ({ useMembersFn, setLoaderModeFn, ...rest }) => {
  const { anomaly, isLoading, members, onChangeSorting, stateSorting, onChangePaging } = useMembersFn({});
  return (
    <MembersProvider value={{ onChangeSorting, sorting: stateSorting }}>
      <Members
        {...rest}
        members={members.data}
        loaderMode={setLoaderModeFn({ isLoading })}
        anomaly={anomaly[SERVICE_NAME]}
        pagination={members.pagination}
        onChangePaging={onChangePaging}
        onChangeSorting={onChangeSorting}
        sorting={stateSorting}
      />
    </MembersProvider>
  );
};

MembersEnhanced.propTypes = {
  useMembersFn: func,
  setLoaderModeFn: func,
};

MembersEnhanced.defaultProps = {
  useMembersFn: useMembers,
  setLoaderModeFn: setLoaderMode,
};

export default MembersEnhanced;
