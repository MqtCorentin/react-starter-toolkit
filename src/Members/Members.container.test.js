import React from 'react';
import { render, act } from '@testing-library/react';
import { MembersEnhanced } from './Members.container';

const defaultProps = {
  header: () => {},
  title: () => {},
  footer: () => {},
  members: [],
  anomaly: null,
  loaderMode: 'none',
  deleteNotification: () => {},
  notifications: [],
};

describe('Component <MembersEnhanced />', () => {
  it('Should render <MembersEnhanced /> with defaultProps', async () => {
    await act(async () => {
      const { asFragment } = render(<MembersEnhanced {...defaultProps} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('Should render <MembersEnhanced /> with one member', () => {
    const useMembersMock = jest.fn(() => ({
      members: [
        {
          _id: '00001',
          type: 'PP',
          firstname: 'Samuel',
          lastname: 'Gomez',
          birthdate: '20/10/1985',
          sexe: 'M',
        },
      ],
      onChangeOrder: () => {},
      sorting: {
        field: '',
        order: 'NONE',
      },
    }));
    act(() => {
      const { asFragment } = render(<MembersEnhanced {...defaultProps} useMembersFn={useMembersMock} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
