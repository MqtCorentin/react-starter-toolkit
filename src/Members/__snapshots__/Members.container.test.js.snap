// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <MembersEnhanced /> Should render <MembersEnhanced /> with defaultProps 1`] = `
<DocumentFragment>
  <div
    class="af-main container"
  >
    <h1
      class="af-title--content"
    >
      Liste des sociétaires
    </h1>
    <div
      style="min-height: 94px; position: relative;"
    >
       
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component <MembersEnhanced /> Should render <MembersEnhanced /> with one member 1`] = `
<DocumentFragment>
  <div
    class="af-main container"
  >
    <h1
      class="af-title--content"
    >
      Liste des sociétaires
    </h1>
    <div
      style="min-height: 94px; position: relative;"
    >
      <table
        class="af-table af-table--members"
      >
        <thead
          class="af-table-thead"
        >
          <tr
            class="af-table__tr"
          >
            <th
              class="af-table__th af-table__th--sortable"
              role="button"
            >
              <button
                class="af-btn af-btn--table-sorting"
                type="button"
              >
                Prénom
                <span
                  class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-sorting"
                />
              </button>
            </th>
            <th
              class="af-table__th af-table__th--sortable"
              role="button"
            >
              <button
                class="af-btn af-btn--table-sorting"
                type="button"
              >
                Nom
                <span
                  class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-sorting"
                />
              </button>
            </th>
            <th
              class="af-table__th af-table__th--sortable"
              role="button"
            >
              <button
                class="af-btn af-btn--table-sorting"
                type="button"
              >
                Date de naissance
                <span
                  class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-sorting"
                />
              </button>
            </th>
            <th
              class="af-table__th"
              role="columnheader"
            >
              Sexe
            </th>
          </tr>
        </thead>
        <tbody
          class="af-table__body"
        >
          <tr
            class="af-table__tr"
          >
            <td
              class="af-table__cell"
            >
              Samuel
            </td>
            <td
              class="af-table__cell"
            >
              Gomez
            </td>
            <td
              class="af-table__cell"
            >
              20/10/1985
            </td>
            <td
              class="af-table__cell"
            >
              M
            </td>
          </tr>
        </tbody>
      </table>
       
    </div>
  </div>
</DocumentFragment>
`;
